### DISCUSSION TITLE
Simpler Quickstart?

### DISCUSSION THREAD

**Discussion by meejah** [ethersync/ethersync#162] [General]
Hello,

I'm trying to experiment with ethersync (on one machine for now), but can't make any "sync-ing" actually happen. I've tried to pare this down to the simplest possible case, with no editors at all involved. This is what I've tried:

- create `playground/` and `/playground/.ethersync`
- `cd playground`
- `ethersync daemon`
- (this spits out a bunch of things, including several lines like `7:42:34  INFO ThreadId(04) ethersync::peer: listen_address=`
- I find one that has `127.0.0.1` in it

In another shell:
- create `playground2/.ethersync`
- `cd playground2/`
- `ethersync daemon --peer /ip4/127.0.0.1/tcp/43631/p2p/12D3KooWNxYW5gnh1xNwWmR41R1DNSV3jb1mwqYK4cFAUN97BaC9 --socket-path /tmp/ethersync2`
- (I have also tried this last step with `--secret <X>` where `<X>` is the 3 words after `7:46:36  INFO ThreadId(08) ethersync::peer: Generated new passphrase:`

In the first shell I do something like `echo "some content" >> foo`. That is, create a file called `foo` with some content.

I expect to see this synchronized over to "playground2" as `playground2/foo`. I do not see any additional log lines in either terminal.

Are my expectations wrong? Should anything at all like this work, or does it only work with editor plugins? (I've tried via the Nix neovim stuff in the quickstart too, but still don't see "synchronizing" -- although I'm an emacs user myself ;) normally).

As for the networking, the unix-socket (i.e. `/tmp/ethersync`) is just for local editors to connect to, right? Other instances would connect via one of the `/ip4/<ip-address>/...` lines right? So e.g. you arrange to forward to port 46631 on interface 127.0.0.1 (in the above example)?

Thanks for any hints!

**Comment by blinry** [comment]
Hey meejah! :) Thanks for trying Ethersync!

Your understanding is pretty correct, and you did everything right! Specifying the `--secret` in the second peer is indeed mandatory to make a connection – this is our authorization method.

The reason why the created file is not synced over is that we haven't merged #133 yet, which will support creating/changing files from the command line. Until then, the "interesting" stuff indeed happens when you use editors.

Two things to help debug:

- Are you getting "Connected to peer/Peer connected" messages in the logs of your two daemons?
- In the Neovim-enabled editor, do you have an `:EthersyncInfo` command? If you open a file in the directory of the first daemon, does it say "Connected to Ethersync daemon"? To open a file in the second directory and see live edits, make sure to set the `ETHERSYNC_SOCKET` variable as explained [here](https://ethersync.github.io/ethersync/editor-plugin-dev-guide.html#starting-two-local-daemons)!

We also recently [merged support](https://github.com/ethersync/ethersync/pull/141) for just opening the same file in multiple editors (using only one daemon). Just make sure to start both editors with the same file content – there's still a subtle bug when the starting buffer contents diverge.

We'd appreciate any thoughts on how to make it easier to get Ethersync to run!

**Comment by meejah** [comment]
Okay, thanks for the confirmation.

Trying again, with 4 shells open. All are in the root of an ethersync checkout. I set `PATH` in all of them so they can see "ethersync". I have a `playground` and `playground2` directory, both with `.ethersync` subdirs.

In shell 1:
- cd `playground`
- run `ethersync demon`
- I see the secret and a bunch of peer addresses logged

In shell 2:
- cd `playground2`
- set `ETHERSYNC_SOCKET` to `/tmp/ethersync2`
- run `ethersync daemon --peer /ip4/127.0.0.1/tcp/40735/p2p/12D3KooWNxYW5gnh1xNwWmR41R1DNSV3jb1mwqYK4cFAUN97BaC9 --secret bank-hair-pen --socket-path /tmp/ethersync2`

So far so good? I do see a `peer connected` in both daemons now.

In shell 3:
- start neovim using Nix (`nix run github:ethersync/ethersync#neovim playground/bar0`)
- After neovim starts up, I see:
```
Connected to Ethersync daemon!
W13: Warning: File "playground/bar0" has been created after editing started                                                   
[O]K, (L)oad File, Load File (a)nd Options:
```
- I say "Okay"

In shell 4:
- set `ETHERSYNC_SOCKET` to `/tmp/ethersync2`
- start neovim using Nix: `nix run github:ethersync/ethersync#neovim playground2/bar0`
- Get the same warning (also say "okay")

Now, **I do see synchronized content** between the two editors!
(So I think it was indeed the "secret" missing in one of my previous attempts).

I've just subscribed to #133 as I think that will make playing with things easier (for me, anyway).

For some concrete improvements to the Quickstart, here is what I remember being confused by:

- it wasn't clear to me that you _need_ to use the editor plugins for anything "interesting" to happen
- it's not at all clear what the secret is for (and the example "joining peer" doesn't use `--secret` in the command-line)
- (for example, I didn't know if it was for peers to authorize localhost connections, or for the daemon to authorize unix-socket editor connections)
- having exact command-lines to paste in would help
- Saying the warning you get is expected / okay (and which option to select) would help
- some screenshots of success might help

Edit: when quitting, you also get warnings that you "haven't written the file". Okay, so then you try to write it and you get a warning that it already has contents ... so explaining that part more fully would be good too (i.e. "just run `q!` to exit without saving, the contents should already be correct on disk) ... right?

**Comment by meejah** [comment]
My _actual_ motivation here is to play with https://github.com/meejah/fowl to get this working across the internet (but without messing with firewall etc settings).

Perhaps once I make a proof-of-concept of that work, I can suggest a "more integrated" wormhole approach to make things easier for peers to connect.

Ideally, this would be via some sort of UI (so probably "via the editor plugins"?) and the humans behind the two peers would just exchange a wormhole code (e.g. the host says "invite a peer!" and gets a code, which they tell to the other human and that client says "join a session: <the code they got told>".

I might need to understand things more fully to know if the above even makes sense :) 

The idea would be that after the above, the two daemons have a persistent, authenticated and encrypted network connection between the peers, and can exchange anything else required to make things work. `fowl` makes it possible to do this without any `ethersync` code changes -- but a more-integrated approach could do more than just forward streams (like `fowl` is doing)

**Comment by blinry** [comment]
Thanks, that's very helpful for us to improve our documentation – which is what we're working on currently anyway! I tried to address some of your suggestions in 7dfd8a8.

You also encountered bug #92, which would be great to fix… (basically, we're creating non-existing files to be able to find out if we should ignore it in Ethersync… not great, but the ignore crate [currently offers no better way](https://github.com/BurntSushi/ripgrep/discussions/2862)).

And yay for your fowl experiments! If you manage to get that to work, definitely let us know! Our other option is to use one of libp2p's [NAT traversal methods](https://docs.libp2p.io/concepts/nat/).

**Comment by meejah** [comment]
Those changes LGTM!

I'll try to express the `fowl` stuff as a PR once I play with it a little more, e.g. add something to the 'pair-programming' use-cases?

**Comment by meejah** [comment]
Hey, I have a draft blog-post talking a bunch about Ethersync -- if you'd like, I can share a "not public" link for you folks to pre-read? (I'm "meejah" on Libera, is one real-time way to contact me .. do you have anything on Matrix or IRC to talk about Ethersync?)

**Reply by blinry** [reply]
Ohh, sounds interesting! We don't have a real-time chat currently, feel free to send to ethersync <ätt> zormit.de. Very curious to read what you wrote!!
